lib/service/service.go

```diff
@@ -2377,6 +2391,80 @@ func (process *TeleportProcess) initDiagnosticService() error {
 	return nil
 }
 
+func (process *TeleportProcess) initTracingService() error {
+	process.log.Info("Initializing tracing provider and exporter.")
+
+	attrs := []attribute.KeyValue{
+		attribute.String(tracing.ProcessIDKey, process.id),
+		attribute.String(tracing.HostnameKey, process.Config.Hostname),
+		attribute.String(tracing.HostIDKey, process.Config.HostUUID),
+	}
+
+	traceConf := tracing.Config{
+		Service:      teleport.ComponentTeleport,
+		Attributes:   attrs,
+		ExporterURL:  process.Config.Tracing.ExporterURL,
+		SamplingRate: process.Config.Tracing.SamplingRate,
+	}
+
+	if len(process.Config.Tracing.KeyPairs) > 0 || len(process.Config.Tracing.CACerts) > 0 {
+		tlsConfig := &tls.Config{}
+		for _, pair := range process.Config.Tracing.KeyPairs {
+			certificate, err := tls.LoadX509KeyPair(pair.Certificate, pair.PrivateKey)
+			if err != nil {
+				return trace.Wrap(err, "failed to read keypair: %+v", err)
+			}
+			tlsConfig.Certificates = append(tlsConfig.Certificates, certificate)
+		}
+
+		pool := x509.NewCertPool()
+		for _, caCertPath := range process.Config.Tracing.CACerts {
+			caCert, err := os.ReadFile(caCertPath)
+			if err != nil {
+				return trace.Wrap(err, "failed to read tracing CA certificate %+v", caCertPath)
+			}
+
+			if !pool.AppendCertsFromPEM(caCert) {
+				return trace.BadParameter("failed to parse tracing CA certificate: %+v", caCertPath)
+			}
+		}
+
+		tlsConfig.ClientCAs = pool
+		tlsConfig.RootCAs = pool
+		traceConf.TLSConfig = tlsConfig
+	}
+
+	log := process.log.WithFields(logrus.Fields{
+		trace.Component: teleport.Component(teleport.ComponentTracing, process.id),
+	})
+
+	process.RegisterFunc("tracing.init", func() error {
+		ctx, cancel := context.WithTimeout(process.ExitContext(), 5*time.Second)
+		defer cancel()
+		provider, err := tracing.NewTraceProvider(ctx, traceConf)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		process.OnExit("tracing.shutdown", func(payload interface{}) {
+			if payload == nil {
+				log.Info("Shutting down immediately.")
+				ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+				defer cancel()
+				warnOnErr(provider.Shutdown(ctx), log)
```

Comment:
FYI the scram shutdown (on SIGINT/SIGTERM) will not wait for services to shutdown, so it's very likely that we won't have the full 5 seconds to perform the provider shutdown - is that going to be a problem?